docs/arch/adr-001.md

```diff
@@ -0,0 +1,38 @@
+# ADR-1: Code Generation
+
+## Context
+
+Serializing and deserializing JSON in Java works best when performing it to/from POJOs.
+While it is possible to serialize into generic `JsonNode` or `Map` objects this results in untyped values that
+must be referenced via String keys, meaning the compiler is limited in what it can check for us.
+
+The classes required for JSON serialization and deserialization are reptitive and error prone to write by hand.
+[Immutables](https://immutables.github.io/) helps to some extent, but there are still many classes and fields to be written.
+
+The [Discord API documentation](https://discordapp.com/developers/docs/intro) is mostly consistent with how it documents
+the JSON structures it returns and accepts.
+
+Code generation is often difficult to use with IDEs.
+Code is often referring to classes that will not be committed, so have to be generated by the IDE.
+
+## Decision
+
+We will use code generation to generate the POJOs to be used for JSON serialization and deserialization.
+
+We will generate classes that make use of Immutables.
+
+## Consequences
+
+The risk of errors when transcribing the Discord API are reduced.
+The specifications can by copy/pasted directly from the documentation and used as the source for code generation.
+
+By using Immutables we reduce the amount of code we must generate.
+For example, Immutables will provide with us with implementations of `equals` and `hashCode`,
+along with a builder for each class.
```

Comment:
Why is this a consequence?

Reply:
As per http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions, consequences describe the results of the decision